source/opt/debug_info_manager.cpp

```diff
@@ -0,0 +1,213 @@
+      context(), SpvOpExtInst, context()->get_type_mgr()->GetVoidTypeId(),
+      result_id,
+      {
+          {spv_operand_type_t::SPV_OPERAND_TYPE_ID,
+           {context()
+                ->get_feature_mgr()
+                ->GetExtInstImportId_OpenCL100DebugInfo()}},
+          {spv_operand_type_t::SPV_OPERAND_TYPE_EXTENSION_INSTRUCTION_NUMBER,
+           {static_cast<uint32_t>(OpenCLDebugInfo100DebugInlinedAt)}},
+          {spv_operand_type_t::SPV_OPERAND_TYPE_LITERAL_INTEGER, {line_number}},
+          {spv_operand_type_t::SPV_OPERAND_TYPE_ID, {scope.GetLexicalScope()}},
+      }));
+  // |scope| already has DebugInlinedAt. We put the existing DebugInlinedAt
+  // into the Inlined operand of this new DebugInlinedAt.
+  if (scope.GetInlinedAt() != kNoInlinedAt) {
+    inlined_at->AddOperand({spv_operand_type_t::SPV_OPERAND_TYPE_RESULT_ID,
+                            {scope.GetInlinedAt()}});
+  }
+  RegisterDbgInst(inlined_at.get());
+  context()->module()->AddExtInstDebugInfo(std::move(inlined_at));
+  return result_id;
+}
+
+Instruction* DebugInfoManager::CreateDebugInfoNone() {
+  uint32_t result_id = context()->TakeNextId();
+  std::unique_ptr<Instruction> dbg_info_none_inst(new Instruction(
+      context(), SpvOpExtInst, context()->get_type_mgr()->GetVoidTypeId(),
+      result_id,
+      {
+          {SPV_OPERAND_TYPE_RESULT_ID,
+           {context()
+                ->get_feature_mgr()
+                ->GetExtInstImportId_OpenCL100DebugInfo()}},
+          {SPV_OPERAND_TYPE_EXTENSION_INSTRUCTION_NUMBER,
+           {static_cast<uint32_t>(OpenCLDebugInfo100DebugInfoNone)}},
+      }));
+
+  RegisterDbgInst(dbg_info_none_inst.get());
+
+  // Add to the front of |ext_inst_debuginfo_|.
+  return context()->module()->ext_inst_debuginfo_begin()->InsertBefore(
+      std::move(dbg_info_none_inst));
+}
+
+Instruction* DebugInfoManager::CloneDebugDeclare(uint32_t orig_var_id,
+                                                 uint32_t new_var_id) {
+  Instruction* dbgdecl = GetDbgDeclareForVar(orig_var_id);
+  if (dbgdecl == nullptr) return nullptr;
+
+  auto* clone = dbgdecl->Clone(context());
+  clone->SetResultId(context()->TakeNextId());
+  clone->GetOperand(kDebugDeclareOperandVariableIndex).words[0] = new_var_id;
+
+  RegisterDbgDeclareForVar(clone);
+  return clone;
+}
+
+Instruction* DebugInfoManager::GetDebugInlinedAt(uint32_t dbg_inlined_at_id) {
+  auto* inlined_at = GetDbgInst(dbg_inlined_at_id);
+  if (inlined_at == nullptr) return nullptr;
+  if (inlined_at->GetOpenCL100DebugOpcode() !=
+      OpenCLDebugInfo100DebugInlinedAt) {
+    return nullptr;
+  }
+  return inlined_at;
+}
+
+Instruction* DebugInfoManager::CloneDebugInlinedAt(uint32_t dbg_inlined_at_id) {
+  auto* inlined_at = GetDebugInlinedAt(dbg_inlined_at_id);
+  if (inlined_at == nullptr) return nullptr;
+  auto* new_inlined_at = inlined_at->Clone(context());
+  new_inlined_at->SetResultId(context()->TakeNextId());
+  RegisterDbgInst(new_inlined_at);
```

Comment:
Should be be registering a debug instruction that is not in the module yet?  Also, to better keep track of ownership (ie who is responsible for deleting the instruction, you should return a unique pointer.  Lastly, what is the reason for not putting it in the module?